crates/sui-source-validation-service/src/lib.rs

```diff
@@ -352,6 +361,35 @@ pub async fn verify_packages(config: &Config, dir: &Path) -> anyhow::Result<Netw
     Ok(lookup)
 }
 
+pub async fn watch_for_upgrades(config: &Config) -> anyhow::Result<()> {
+    let mut watch_ids = ArrayParams::new();
+    for s in &config.packages {
+        let packages = match s {
+            PackageSources::Repository(RepositorySource { packages, .. }) => packages,
+            PackageSources::Directory(DirectorySource { packages, .. }) => packages,
+        };
+        for p in packages {
+            if let Some(id) = p.watch {
+                watch_ids.insert(TransactionFilter::ChangedObject(id))?
+            }
+        }
+    }
+
+    let client: WsClient = WsClientBuilder::default().build(LOCALNET_WS_URL).await?;
+    let mut subscription: Subscription<SuiTransactionBlockEffects> = client
+        .subscribe(
+            "suix_subscribeTransaction",
+            watch_ids,
+            "suix_unsubscribeTransaction",
+        )
+        .await?;
+
+    info!("Listening for upgrades...");
+    let result = subscription.next().await;
+    info!("Saw upgrade txn: {:?}", result);
```

Comment:
I did wrap this in a `for { ... }` loop to test things locally. I took the loop away to simplify the test for now and check in the basic necessities, and will revisit in follow up.

Reply:
Note that the complicated part about using the subscription APIs is dealing with connection drops: You will need to keep track of the latest transaction you have processed, and in case you have a connection drop, when you resume, you need to issue a call to `suix_queryTransactionBlocks` to get the transactions you may have missed in the meanwhile.

Reply:
Right, thanks for flagging! I have a couple of ideas around robustness that hopefully doesn't make this more painful than it has to be. E.g., maybe there is a way to kick the server or restart subscription for connection drops that doesn't leave big room for missing transactions.

I think the concern around missing transactions is more severe for framework upgrades (if we miss a framework upgrade and fail to reverify, and show source for a previous version--here we might have to track protocol versions). That said, I feel like we have alternative ways and fallbacks to ensure we see framework upgrades (in which case, e.g., restarting / kicking the main server loop will pick up the latest changes).

For third-part/external crates, it's a bit less severe because if we miss a transaction we just won't display source for that package (so at least we are not showing false things).

Reply:
The process for resuming is fairly standard, as long as you keep track of the latest transaction you've processed from the subscription API: 

- When you restart the server or the subscription connection, after you get the first response back from the subscription API, you will get a response containing a cursor for the start of the subscription response.
- You can then issue a `suix_queryTransactionBlocks` request to get the transactions between the one that you last processed and the new one you are about to process.
- Depending on whether your processing is sensitive to order or not, you can then either choose to wait to process your subscription transactions until you get the query response back and process it, or you can do the two in parallel (but this risks inconsistencies).

Reply:
ah, I missed the potential of using a cursor, cheers!